databroker/broker.py

```diff
@@ -880,3 +902,161 @@ def _munge_time(t, timezone):
+                                external='FILESTORE:')
+
+                elif name == 'event':
+                    # We need a selectively deeper copy since we will mutate
+                    # fs_doc['data'].
+                    fs_doc['data'] = dict(fs_doc['data'])
+                    # The writer writes data to an external file, creates a
+                    # datum record in the filestore database, and return that
+                    # datum_id. We modify fs_doc in place, replacing the data
+                    # values with that datum_id.
+                    for data_key, writer in writers.items():
+                        fs_uid = writer.write(fs_doc['data'][data_key])
+                        fs_doc['data'][data_key] = fs_uid
+
+                    doc.update(
+                        filled={k: True for k in external_writers.keys()})
+
+                elif name == 'stop':
+                    for data_key, writer in list(writers.items()):
+                        writer.close()
+                        writers.pop(data_key)
+
+                # The mutated fs_doc is inserted into metadatastore.
+                fs_doc.pop('filled', None)
+                fs_doc.pop('_name', None)
+                db.mds.insert(name, fs_doc)
+
+                # The pristine doc is yielded.
+                yield name, doc
+
+        return wrapped_f
+
+    return wrap
+
+
+def event_map(stream_name, data_keys, provenance):
+    """
+    Map a function onto each event in a stream.
+
+    Parameters
+    ----------
+    stream_name : string
+        e.g., 'primary' or 'baseline'
+    data_keys : dict
+        key(s) in event['data'] to apply function to (e.g., 'image') mapped to
+        a dict with any updates to the data key value
+        (e.g., {'shape': [10, 10]}). In the simple case where the shape,
+        datatype, etc. are unchanged, the dict is just empty:
+        ``{'image': {}}`` and the original metadata passed through.
+    proveance : dict
+        metadata about this operation
+    """
+    def outer(f):
+        def inner(stream):
+            run_start_uid = None
+            descriptor_uid = None
+            for name, doc in stream:
+                if name == 'start':
+                    run_start_uid = str(uuid.uuid4())
+                    new_start_doc = dict(uid=run_start_uid,
+                                         time=time.time(),
+                                         parents=[doc['uid']],
+                                         provenance=provenance)
+                    yield 'start', new_start_doc
+
+                elif name == 'descriptor' and doc.get('name') == stream_name:
+                    if run_start_uid is None:
+                        raise RuntimeError("Received EventDescriptor before "
+                                           "RunStart.")
+                    descriptor_uid = doc_or_uid_to_uid(doc)
+                    new_data_keys = dict(doc['data_keys'])
```

Comment:
Should we be eating all the keys from the prior run, even the ones we aren't going to transform?